core/shared/src/main/scala/io/circe/Decoder.scala

```diff
@@ -389,17 +389,15 @@ object Decoder extends TupleDecoders with LowPriorityDecoders {
    * @group Decoding
    */
   implicit def decodeOption[A](implicit d: Decoder[A]): Decoder[Option[A]] =
-    withReattempt { a =>
-      a.success.fold[Result[Option[A]]](Xor.right(None)) { valid =>
-        if (valid.focus.isNull) Xor.right(None) else d(valid).fold[Result[Option[A]]](
-          df =>
-            df.history.headOption.fold[Result[Option[A]]](
-              Xor.right(None)
-            )(_ => Xor.left(df)),
-          a => Xor.right(Some(a))
-        )
-      }
-    }
+    withReattempt(c =>
+      if (c.succeeded) {
+        if (c.any.focus.isNull) Xor.right(None) else d(c.any) match {
```

Comment:
Low hanging fruit (-8 bytes) is `Xor.right(None)` - I think we could lift this one into a `val` and reuse it whenever possible (pretty sure there are a lot of places).